Navigation: Applying styles & behavior for folded menu, including mobile

joelclimbsthings opened the pull request:
Fixes #5326 

Updating styles and behaviour governing new navigation feature, including mobile. This should align with the following acceptance criteria: 

- A closed menu will appear as a square with the site logo of size equal to the wc-admin header.
- With the navigation closed, wc-admin or embedded pages occupy the entire screen width.
- On screens `$break-large: 960px` or below toggling the navigation pushes or overlaps the content, whichever is better.
- At the same breakpoint as above, the Navigation is collapsed by default.

I've also added a listener on the `orientationchange` event to handle the case of the layout for a device switching from landscape to portrait, in which case it will hide the navigation. This change is probably debatable usability wise, so I'm glad to remove it if desired.

### Screenshots
**Folded Before:** 
<img width="459" alt="Screen Shot 2020-10-23 at 3 36 04 PM" src="https://user-images.githubusercontent.com/1922453/96949927-91531500-1545-11eb-8743-66b72cde2998.png">

**Folded After:**
![image](https://user-images.githubusercontent.com/444632/97493543-57598880-1922-11eb-947f-66dd059b989f.png)

**Open on Desktop:**
![image](https://user-images.githubusercontent.com/444632/97493655-866ffa00-1922-11eb-86c3-fb52f5a11708.png)

**Open on Mobile:**
![image](https://user-images.githubusercontent.com/444632/97493734-9f78ab00-1922-11eb-8a24-8919001b83ac.png)


### Detailed test instructions:

- Navigate to WooCommerce pages with navigation feature on.
- Click logo in top-left to fold navigation.
- Ensure your screen width is greater than 960px.
- Observe that it now appears as the "Folded After" screenshot, and does **not** overlap content.
- Shrink viewport to less than 960px.
- With the navigation open, it should appear the same but now overlay the content.
- Refresh the page, and ensure that the  navigation collapses by default (while still with a small viewport).
- You can use Chrome or Firefox dev tools to test a `orientationchange` event to ensure that the navigation collapses when switching to a view smaller than `960px`.

Commit:
Adding base styles for collapsed nav menu.

Commit:
Handling varying behavior on mobile-sized screens, with the menu overlaying content instead of shrinking the content area.

Commit:
Folding the menu by default on screen sizes less than 960px, and if that condition is met after an orientation change

joshuatf commented on client/navigation/stylesheets/index.scss:
```diff
@@ -17,28 +17,53 @@
+		height: 100%;
+		box-sizing: border-box;
+		background-color: $gray-900;
+		z-index: 1100;
```
Out of curiosity, what was the reason behind picking `1100`?  If it's higher than another, a comment explaining why it should be > n may be helpful to someone working on this in the future.

joshuatf commented on client/navigation/components/header/index.js:
```diff
@@ -15,6 +16,22 @@ const Header = () => {
+
+		window.addEventListener(
+			'orientationchange',
+			( e ) => foldOnMobile( e.target.screen.availWidth ),
```
Should this use `width` instead of `availWidth`?

psealock replied:
`document.body.clientWidth` and `screen.availWidth` give me differing results if I have Chrome console open. Should we stick with one or the other?

joshuatf commented on client/navigation/components/header/index.js:
```diff
@@ -15,6 +16,22 @@ const Header = () => {
+		foldOnMobile( document.body.clientWidth );
+
+		window.addEventListener(
+			'orientationchange',
```
Should this happen on resize as well?  The way the current WP menu changes also occurs on resize, but I can see a case being made for that not happening.

psealock replied:
I'm ok with leaving out the resize eventListener. It feels like something only devs do when doing code review 😄

joshuatf reviewed (commented):
Nice work, Joel!  This is testing well for me, just left a few questions.

Also curious if this is the expected mobile interaction you described @jameskoster.  You mentioned "sliding"- should this animate at all from folded to unfolded?

psealock reviewed (commented):
Thanks @joelclimbsthings, this is testing nicely for me. I added a quick comment on top of Josh's and I did notice one small thing.

When you scroll up, the header bar gets a drop shadow. Can you add it to the Navigation button as well?

![Screen Shot 2020-10-29 at 3 09 23 PM](https://user-images.githubusercontent.com/1922453/97517615-67449b80-19fa-11eb-98f2-d1af4f955531.png)

psealock commented:
One more thing I noticed: It would be great to prevent the animation on mount so it doesn't happen when folding and unfolding the menu. That might be good for a follow up because it could involve a Gutenberg PR.

Commit:
Adding comment to zIndex style change for clarity

Commit:
Refactoring logic from Header/index.js to flag if user has scrolled to apply shadow to folded navigation as well.

joelclimbsthings commented:
> When you scroll up, the header bar gets a drop shadow. Can you add it to the Navigation button as well?

Hey @psealock , I've just made a commit to address this. It's slightly more complex than expected since this shadow is triggered via  class that can't be referenced by our relevant styles. I ended up refactoring the logic that attaches the `is-scrolled` class that creates this shadow and implementing it on our component as well. 

I didn't see any other reusuable hooks in the project, so just let me know if another approach is preferred. It's possible that a more performant solution would be possible (since this one actually attaches two scroll events I believe), but these components don't have a common ancestor or anything to attach a `Context`. No redux, of course, but perhaps `@wordpress/data` can be used for something like this? Haven't dived into that particular rabbit hole just yet.

Screenshot:
![image](https://user-images.githubusercontent.com/444632/97760502-860d6580-1ac0-11eb-9c63-6ad4236324b3.png)

Also, I actually found a bug that was stemming from the new navigation. The shadow is actually supposed to disappear when you scroll back to the top, but this had been broken. Now fixed as well.

psealock commented on client/hooks/useIsScrolled.js:
```diff
@@ -0,0 +1,29 @@
+			rafHandle.current = window.requestAnimationFrame(
+				updateIsScrolled
+			);
+		};
```
I'm not sure whats going on, but the contents of the `scrollListener` don't get called on scroll so its not updating on scroll. The `is-scrolled` class is working well if I apply it manually though.

joelclimbsthings replied:
Strange, it was working reliably for me on Friday. I'll take another look. You're just Firefox or Chrome?

joelclimbsthings replied:
Yeah, can't reproduce this issue @psealock . Which page/etc are you on?

psealock replied:
Ok, it appears to be related to this error being displayed:

<img width="1027" alt="Screen Shot 2020-11-03 at 9 41 14 AM" src="https://user-images.githubusercontent.com/1922453/97917094-cfa7ca00-1db8-11eb-9c72-f03c84ff5bed.png">

When I comment out `define( 'WP_DEBUG', true );` to turn that off, it works great. I think the header uses some position CSS to go over the error message, but I have no idea why that would interfere with the scroll eventListener not being called.

joelclimbsthings replied:
Ah, right. Not sure why it's conflict here either, but I've had my `WP_DEBUG` set to false since my page is overrun with deprecation warnings otherwise.

psealock reviewed (commented):
Nice catch on the scroll issue @joelclimbsthings. Something isn't testing quite right for me because the `is-scrolled` class never gets applied.

> It's possible that a more performant solution would be possible (since this one actually attaches two scroll events I believe)

The only thing I can think of is some kind of cache or flag to keep track if the listener has already been added to the window. I do see two `scroll` listeners added as a result of `useIsScrolled` in dev tools. It might be good to merge this and see if a better way exists in a follow up if you'd like.

joshuatf commented on client/navigation/stylesheets/index.scss:
```diff
@@ -17,28 +17,59 @@
+
+			.woocommerce-navigation-header {
+				&.is-scrolled {
+					box-shadow: 0 8px 8px 0 rgba(85, 93, 102, 0.3);
```
A variable for this might be good so we don't get out of sync with the header should it ever change.

joelclimbsthings replied:
Good call. Done :+1:

joshuatf reviewed (commented):
Testing well for me.  The `is-scrolled` class did get added, so this looks good on my end unless @psealock is able to reproduce this issue.

joelclimbsthings committed:
Refactoring styles in navigation feature to allow them to be loaded properly with webpack (with imports/etc).

joelclimbsthings committed:
Adding central variable for header box-shadow when scrolling

joelclimbsthings commented:
I believe I've addressed all of the feedback at this point, with the potential of a couple follow-up issues regarding performance and a possible GB PR.

psealock reviewed (approved):
LGTM, although that is weird about the conflict with the warning. Maybe there is a fix to look into after this.

Thanks @joelclimbsthings
